packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level1/coords.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (C) 2019 Matthew Jennings
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+"""A suite of functions for handling DICOM coordinates"""
+
+import numpy as np
+
+from pymedphys_utilities.libutils import get_imports
+IMPORTS = get_imports(globals())
+
+
+def coords_from_xyz_axes(xyz_axes):
+    """Converts a set of x, y and z axes of a regular grid (e.g. a DICOM
+    pixel array) into an array of three grids whose voxels correspond to
+    and contain the `x`, `y`, and `z` coordinates of the original grid.
+
+    Parameters
+    ----------
+    xyz_axes : tuple
+        A tuple containing three `ndarrays` corresponding to the `x`,
+        `y` and `z` axes of a given 3D grid - usually a DICOM dataset's
+        pixel array.
+
+    Returns
+    -------
+    coords : ndarray
+        An array containing three grids consisting of the `x`, 'y` and
+        `z` coordinates of the corresponding grid (e.g. DICOM dataset's
+        pixel array) from which the original axes were extracted.
+    """
+    ZZ, YY, XX = np.meshgrid(
+        xyz_axes[2], xyz_axes[1], xyz_axes[0], indexing='ij')
```

Comment:
I seem to remember this order here is important as in it decides which axis corresponds to which coordinate value. What's the responding for the reverse order? Not against it, just interested why the choice.

Reply:
This is the order required for the following to be true:

The coordinate value at `coords[axis, i, j, k]` corresponds to the dose value at `dose[i, j, k]`.

When you read `grid = ds.pixel_array`, the index order is `grid[slice, row, column]`, which corresponds to `grid[z, y, x]` in the DICOM coordinate system.

Reply:
I think this was all relatively subconscious, but in your older functions (where you were extracting PDDs, etc.) you were working in the IEC fixed coordinate system. In the IEC fixed coordinate system, `grid[slice, row, column]` corresponds to `grid[y, z, x]`. You also seemed to prefer the order `grid[row, column, slice]` (corresponding to  `grid[y, x, z]`), hence the transpose [here](https://github.com/pymedphys/pymedphys/blob/4d200898f678d2306c49b517ab9d8700270063cc/packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level2/dose.py#L75).

Finally, when you were working with `coords` (now called "axes") and `dose` together, you tended to reorder the coords tuple to `(y, x, z)` to match your adjusted dose grid. E.g. [here](https://github.com/pymedphys/pymedphys/blob/4d200898f678d2306c49b517ab9d8700270063cc/packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level2/dose.py#L425).

Reply:
Later we agreed to keep the dose grid unaltered (things got messy by both transposing the grid, AND reordered coords, plus other aesthetic reasons besides). So we wrote some coords extraction functions for different coordinate systems to handle cases where they might be more intuitive (like water tank applications).

Reply:
`y, x, z` corresponds to `i, j, k` index order, which is preferred for linear algebra. Consequently, it appears to have made it into standard programming practice. It's just that for our application, axis order is probably more important than index order (when dealing with coordinates, physicists are going to much prefer xyz, ijk is unlikely to enter their minds). Moreover, I'd say medical physicists are probably more used to a kji index order *anyway*, since we're used to CT/volumetric applications where we typically *select slice first*.

Reply:
I'm all okay with this :). It'll just take me a bit to get used to I think.